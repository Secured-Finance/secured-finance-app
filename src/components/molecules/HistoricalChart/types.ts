import { ISeriesApi } from 'lightweight-charts';
import { HistoricalDataIntervals } from 'src/types';

export interface ITradingData {
    time: string;
    open: number;
    high: number;
    low: number;
    close: number;
    vol: number;
}

export interface HistoricalChartProps {
    data: ITradingData[];
    timeScale?: HistoricalDataIntervals;
}

export type TSeries = ISeriesApi<
    'Line' | 'Area' | 'Histogram' | 'Candlestick' | 'Bar'
>;
