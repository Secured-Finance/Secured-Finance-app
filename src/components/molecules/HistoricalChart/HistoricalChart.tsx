import { getUTCMonthYear } from '@secured-finance/sf-core';
import clsx from 'clsx';
import {
    CandlestickData,
    HistogramData,
    IChartApi,
    ISeriesApi,
    LogicalRange,
    MouseEventParams,
    Time,
    UTCTimestamp,
    WhitespaceData,
} from 'lightweight-charts';
import { useCallback, useEffect, useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import { selectLandingOrderForm } from 'src/store/landingOrderForm';
import { RootState } from 'src/store/types';
import { createCandlestickChart, createVolumeChart } from 'src/utils/charts';

export interface ITradingData {
    time: number;
    open: number;
    high: number;
    low: number;
    close: number;
    vol: number;
}

export interface HistoricalChartProps {
    data: ITradingData[];
    className?: string;
}

type TSeries = ISeriesApi<
    'Line' | 'Area' | 'Histogram' | 'Candlestick' | 'Bar'
>;

function getCrosshairDataPoint(series: TSeries, param: MouseEventParams) {
    if (!param.time) {
        return null;
    }
    const dataPoint = param.seriesData.get(series);
    return dataPoint || null;
}

function syncCrosshair(
    chart: IChartApi,
    series: TSeries,
    dataPoint:
        | WhitespaceData<Time>
        | HistogramData<Time>
        | CandlestickData<Time>
        | null
) {
    if (dataPoint) {
        chart.setCrosshairPosition(
            (dataPoint as HistogramData<Time>).value,
            dataPoint.time,
            series
        );
        return;
    }
    chart.clearCrosshairPosition();
}

export function HistoricalChart({ data, className }: HistoricalChartProps) {
    const chartContainerRef = useRef<HTMLDivElement>(null);
    const secondContainerRef = useRef<HTMLDivElement>(null);
    const [hoverTime, setHoverTime] = useState('');
    const { currency, maturity } = useSelector((state: RootState) =>
        selectLandingOrderForm(state.landingOrderForm)
    );
    const prettyMaturity = getUTCMonthYear(+maturity);
    const VOLUME_KEY_NAME = `Vol(${currency} ${prettyMaturity})`;

    const [legendData, setLegendData] = useState({
        O: '',
        H: '',
        L: '',
        C: '',
        VOLUME_KEY_NAME: '',
        Change: '',
    });

    const setupCharts = useCallback(
        (candlestickSeries: TSeries, volumeSeries: TSeries) => {
            volumeSeries.applyOptions({ baseLineVisible: false });

            const candleData = data.map(item => ({
                time: Math.floor(item.time) as UTCTimestamp,
                open: Number(item.open),
                high: Number(item.high),
                low: Number(item.low),
                close: Number(item.close),
            }));

            const volumeData = data.map(item => ({
                time: Math.floor(item.time) as UTCTimestamp,
                value: Number(item.vol),
                color: item.open > item.close ? '#FF9FAE' : '#09A8B7',
            }));

            candlestickSeries.setData(candleData);
            volumeSeries.setData(volumeData);
        },
        [data]
    );

    const subscribeToChartEvents = (
        candleStickChart: IChartApi,
        volumeChart: IChartApi,
        candlestickSeries: TSeries,
        volumeSeries: TSeries
    ) => {
        const updateLegendData = (param: MouseEventParams<Time>) => {
            if (!param.seriesData.size) return;

            let index = 0;
            if (param.time) {
                index = candlestickSeries
                    .data()
                    .findIndex(
                        (item: Record<'time', Time>) => item.time === param.time
                    );
            }

            const candleData = candlestickSeries.dataByIndex(index);
            const volumeData = volumeSeries.dataByIndex(index);
            const mergeData = {
                ...candleData,
                ...volumeData,
            } as Omit<ITradingData, 'vol'> & { value: number };

            const date = new Date(Number(candleData?.time) * 1000);
            const formattedDate =
                date.getFullYear() +
                '/' +
                ('0' + (date.getMonth() + 1)).slice(-2) +
                '/' +
                ('0' + date.getDate()).slice(-2);
            setHoverTime(formattedDate);

            setLegendData({
                O: `${mergeData?.open}`,
                H: `${mergeData?.high}`,
                L: `${mergeData?.low}`,
                C: `${mergeData?.close}`,
                VOLUME_KEY_NAME: mergeData?.value?.toFixed(2),
                Change: `${(
                    ((mergeData?.close - mergeData?.open) / mergeData?.open) *
                    100
                ).toFixed(2)}%`,
            });
        };

        candleStickChart.subscribeCrosshairMove(function (param) {
            const dataPoint = getCrosshairDataPoint(candlestickSeries, param);
            syncCrosshair(volumeChart, volumeSeries, dataPoint);
            updateLegendData(param);
        });

        volumeChart.subscribeCrosshairMove(function (param) {
            const dataPoint = getCrosshairDataPoint(volumeSeries, param);
            syncCrosshair(candleStickChart, candlestickSeries, dataPoint);
            updateLegendData(param);
        });

        candleStickChart
            .timeScale()
            .subscribeVisibleLogicalRangeChange(range => {
                volumeChart
                    .timeScale()
                    .setVisibleLogicalRange(range as LogicalRange);
            });

        volumeChart.timeScale().subscribeVisibleLogicalRangeChange(range => {
            candleStickChart
                .timeScale()
                .setVisibleLogicalRange(range as LogicalRange);
        });
    };

    useEffect(() => {
        if (!chartContainerRef.current || !secondContainerRef.current) return;
        const { candlestickSeries, chart: candleStickChart } =
            createCandlestickChart(chartContainerRef.current);

        const { volumeSeries, chart: volumeChart } = createVolumeChart(
            secondContainerRef.current
        );

        setupCharts(candlestickSeries, volumeSeries);

        subscribeToChartEvents(
            candleStickChart,
            volumeChart,
            candlestickSeries,
            volumeSeries
        );

        volumeChart.timeScale();
        candleStickChart.timeScale();

        const updateTitleData = (
            param: MouseEventParams<Time>,
            sourceChart: string
        ) => {
            if (!param.seriesData.size) return;

            let index = 0;
            if (param.time) {
                index = candlestickSeries
                    .data()
                    .findIndex(
                        (item: Record<'time', Time>) => item.time === param.time
                    );
            }

            const candleData = candlestickSeries.dataByIndex(index);
            const volumeData = volumeSeries.dataByIndex(index);
            if (sourceChart === 'candlestick') {
                syncCrosshair(volumeChart, volumeSeries, candleData);
            } else if (sourceChart === 'volume') {
                syncCrosshair(candleStickChart, candlestickSeries, volumeData);
            }
            const mergeData = {
                ...candleData,
                ...volumeData,
            } as Omit<ITradingData, 'vol'> & { value: number };

            const date = new Date(Number(candleData?.time) * 1000);
            const formattedDate =
                date.getFullYear() +
                '/' +
                ('0' + (date.getMonth() + 1)).slice(-2) +
                '/' +
                ('0' + date.getDate()).slice(-2);
            setHoverTime(formattedDate);

            setLegendData({
                O: `${mergeData?.open}`,
                H: `${mergeData?.high}`,
                L: `${mergeData?.low}`,
                C: `${mergeData?.close}`,
                VOLUME_KEY_NAME: mergeData?.value?.toFixed(2),
                Change: `${(
                    ((mergeData?.close - mergeData?.open) / mergeData?.open) *
                    100
                ).toFixed(2)}%`,
            });
        };

        candleStickChart.subscribeCrosshairMove(function (param) {
            updateTitleData(param, 'candlestick');
        });

        volumeChart.subscribeCrosshairMove(function (param) {
            updateTitleData(param, 'volume');
        });

        return () => {
            candleStickChart.remove();
            volumeChart.remove();
        };
    }, [data, setupCharts]);

    const titleOfChartClass =
        'z-10 flex gap-4 text-2xs text-neutral-4 font-medium leading-4 pt-[0.375rem] px-4';
    return (
        <div className={clsx(className, 'bg-neutral-900 pt-[0.625rem]')}>
            <div className={clsx(titleOfChartClass)}>
                <div>{hoverTime}</div>
                {Object.entries(legendData)
                    .filter(([key, _]) =>
                        ['O', 'H', 'L', 'C', 'Change'].includes(key)
                    )
                    .map(([key, value]) => {
                        return (
                            <div key={key} className='flex gap-1'>
                                <div className='text-neutral-4'>{key}</div>
                                <div
                                    className={clsx('font-normal', {
                                        'text-[#F5F6FF]': key !== 'Change',
                                        'text-nebulaTeal':
                                            key === 'Change' &&
                                            !value.includes('-'),
                                        'text-galacticOrange':
                                            key === 'Change' &&
                                            value.includes('-'),
                                    })}
                                >
                                    {value}
                                </div>
                            </div>
                        );
                    })}
            </div>
            <div
                ref={chartContainerRef}
                className='relative h-[260px] w-full'
            ></div>
            <div className={clsx(titleOfChartClass)}>
                <div>
                    <span>{VOLUME_KEY_NAME}:</span>
                    <span className='ml-[0.375rem] text-[#FF9FAE]'>
                        {(Number(legendData.VOLUME_KEY_NAME) / 1000).toFixed(3)}
                        K
                    </span>
                </div>
                <div>
                    <span>Vol(USDT):</span>
                    <span className='text-[#FF9FAE]'>
                        {/* TODO: Add vol(USDT) */}
                    </span>
                </div>
            </div>

            <div
                ref={secondContainerRef}
                className={clsx(className, 'relative h-[170px] w-full')}
            ></div>
        </div>
    );
}
