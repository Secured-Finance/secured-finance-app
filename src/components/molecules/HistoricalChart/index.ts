export * from './HistoricalChart';
export * from './types';
