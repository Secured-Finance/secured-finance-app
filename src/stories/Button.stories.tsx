import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Button, IButton } from 'src/components/new/Button';

export default {
    title: 'Components/Button',
    component: Button,
    argTypes: { onClick: { action: 'clicked' } },
} as Meta;

const Template: Story<IButton> = args => <Button {...args}>Button</Button>;

export const Default = Template.bind({});
Default.args = {
    outline: false,
};

export const Outlined = Template.bind({});
Outlined.args = {
    outline: true,
};
