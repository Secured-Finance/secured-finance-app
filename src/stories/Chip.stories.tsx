import React from 'react';
import { Story, Meta } from '@storybook/react';
import cm from './index.module.scss';

import { ChipButton, IChipButton } from 'src/components/new/Chip/ChipButton';
import { ChipLabel, IChipLabel } from 'src/components/new/Chip/ChipLabel';

export default {
    title: 'Components/Chips',
    argTypes: { onClick: { action: 'clicked' } },
} as Meta;

export const ChipButtons: Story<IChipButton> = () => (
    <div className={cm.elementsContainer}>
        <ChipButton>Button</ChipButton>
        <ChipButton accent={'purple'}>Button</ChipButton>
        <ChipButton accent={'red'}>Button</ChipButton>
    </div>
);

export const ChipLabels: Story<IChipLabel> = () => (
    <div className={cm.elementsContainer}>
        <ChipLabel label={'Registered'} />
        <ChipLabel label={'Working'} accent={'green'} />
        <ChipLabel label={'Terminated'} accent={'red'} />
        <ChipLabel label={'Closed'} accent={'grey'} />
        <ChipLabel label={'Past Due'} accent={'turquoise'} />
        <ChipLabel label={'Due'} accent={'blue'} />
    </div>
);
