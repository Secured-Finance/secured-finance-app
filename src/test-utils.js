import { render as rtlRender } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from 'src/store/reducers';

function render(
    ui,
    {
        initialState,
        store = createStore(reducer, initialState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {
        return <Provider store={store}>{children}</Provider>;
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { render };
