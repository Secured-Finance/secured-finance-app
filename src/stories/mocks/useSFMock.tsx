import { Currency } from '@secured-finance/sf-core';
import { BigNumber } from 'ethers';
import * as jest from 'jest-mock';
import { CurrencySymbol, getCurrencyMapAsList } from 'src/utils';
import { collateralBook80, dec24Fixture } from './fixtures';
import {
    dec22Fixture,
    mar23Fixture,
    ethBytes32,
    efilBytes32,
} from './fixtures';

export const mockUseSF = () => {
    const mockSecuredFinance = {
        placeOrder: jest.fn(),
        placePreOrder: jest.fn(),
        getBorrowUnitPrices: jest.fn(() =>
            Promise.resolve([
                BigNumber.from(9687),
                BigNumber.from(9685),
                BigNumber.from(9679),
                BigNumber.from(9674),
                BigNumber.from(9653),
                BigNumber.from(9643),
                BigNumber.from(9627),
                BigNumber.from(9617),
            ])
        ),
        getLendUnitPrices: jest.fn(() =>
            Promise.resolve([
                BigNumber.from(9685),
                BigNumber.from(9683),
                BigNumber.from(9677),
                BigNumber.from(9672),
                BigNumber.from(9651),
                BigNumber.from(9641),
                BigNumber.from(9625),
                BigNumber.from(9615),
            ])
        ),
        getMidUnitPrices: jest
            .fn<Promise<BigNumber[]>, Currency[]>()
            .mockImplementation((ccy: Currency) => {
                switch (ccy.symbol) {
                    case CurrencySymbol.ETH:
                        return Promise.resolve([
                            BigNumber.from(9686),
                            BigNumber.from(9684),
                            BigNumber.from(9678),
                            BigNumber.from(9673),
                            BigNumber.from(9652),
                            BigNumber.from(9642),
                            BigNumber.from(9626),
                            BigNumber.from(9616),
                        ]);
                    case CurrencySymbol.EFIL:
                        return Promise.resolve([
                            BigNumber.from(9586),
                            BigNumber.from(9584),
                            BigNumber.from(9578),
                            BigNumber.from(9573),
                            BigNumber.from(9552),
                            BigNumber.from(9542),
                            BigNumber.from(9526),
                            BigNumber.from(9516),
                        ]);
                    case CurrencySymbol.USDC:
                        return Promise.resolve([
                            BigNumber.from(9486),
                            BigNumber.from(9484),
                            BigNumber.from(9478),
                            BigNumber.from(9473),
                            BigNumber.from(9452),
                            BigNumber.from(9442),
                            BigNumber.from(9426),
                            BigNumber.from(9416),
                        ]);
                    case CurrencySymbol.WBTC:
                        return Promise.resolve([
                            BigNumber.from(9386),
                            BigNumber.from(9384),
                            BigNumber.from(9378),
                            BigNumber.from(9373),
                            BigNumber.from(9352),
                            BigNumber.from(9342),
                            BigNumber.from(9326),
                            BigNumber.from(9316),
                        ]);
                    default:
                        throw new Error('Not implemented');
                }
            }),
        getCollateralBook: jest.fn(() =>
            Promise.resolve({
                collateral: {
                    ...collateralBook80.collateral,
                    ...collateralBook80.nonCollateral,
                },
                collateralCoverage: collateralBook80.coverage,
            })
        ),

        getLendingMarket: jest.fn((_, maturity: number) => {
            if (maturity === dec24Fixture.toNumber()) {
                return Promise.resolve({
                    isItayosePeriod: jest.fn(() => Promise.resolve(false)),
                    isOpened: jest.fn(() => Promise.resolve(false)),
                    isPreOrderPeriod: jest.fn(() => Promise.resolve(true)),
                });
            } else {
                return Promise.resolve({
                    isItayosePeriod: jest.fn(() => Promise.resolve(false)),
                    isOpened: jest.fn(() => Promise.resolve(true)),
                    isPreOrderPeriod: jest.fn(() => Promise.resolve(false)),
                });
            }
        }),

        getMaturities: jest.fn(() =>
            Promise.resolve([
                BigNumber.from('1000'),
                BigNumber.from('2000'),
                BigNumber.from('3000'),
                BigNumber.from('4000'),
                BigNumber.from('5000'),
            ])
        ),

        getLendingMarkets: jest.fn(() =>
            Promise.resolve([
                {
                    midRate: 100,
                    lendRate: 200,
                    borrowRate: 300,
                    maturity: 1000,
                    name: 'ETH-1000',
                    utcOpeningDate: 1620000000,
                },
                {
                    midRate: 100,
                    lendRate: 200,
                    borrowRate: 300,
                    maturity: 2000,
                    name: 'ETH-2000',
                    utcOpeningDate: 1720000000,
                },
            ])
        ),

        depositCollateral: jest.fn(() =>
            Promise.resolve({
                wait: jest.fn(() => Promise.resolve({ blockNumber: 123 })),
                to: '0xb98bd7c7f656290071e52d1aa617d9cb4467fd6d',
            })
        ),

        withdrawCollateral: jest.fn<
            Promise<{
                wait: () => Promise<{
                    blockNumber: number | undefined;
                }>;
            }>,
            []
        >(() =>
            Promise.resolve({
                wait: () => Promise.resolve({ blockNumber: 123 }),
            })
        ),

        getBorrowOrderBook: jest.fn(() =>
            Promise.resolve({
                unitPrices: [
                    BigNumber.from(9690),
                    BigNumber.from(9687),
                    BigNumber.from(9685),
                    BigNumber.from(9679),
                    BigNumber.from(9674),
                ],
                amounts: [
                    BigNumber.from('43000000000000000000000'),
                    BigNumber.from('23000000000000000000000'),
                    BigNumber.from('15000000000000000000000'),
                    BigNumber.from('12000000000000000000000'),
                    BigNumber.from('1800000000000000000000'),
                ],
                quantities: [
                    BigNumber.from('1000'),
                    BigNumber.from('2000'),
                    BigNumber.from('3000'),
                    BigNumber.from('4000'),
                    BigNumber.from('5000'),
                ],
            })
        ),

        getLendOrderBook: jest.fn(() =>
            Promise.resolve({
                unitPrices: [
                    BigNumber.from(9690),
                    BigNumber.from(9687),
                    BigNumber.from(9685),
                    BigNumber.from(9679),
                    BigNumber.from(9674),
                ],
                amounts: [
                    BigNumber.from('43000000000000000000000'),
                    BigNumber.from('23000000000000000000000'),
                    BigNumber.from('15000000000000000000000'),
                    BigNumber.from('12000000000000000000000'),
                    BigNumber.from('1800000000000000000000'),
                ],
                quantities: [
                    BigNumber.from('1000'),
                    BigNumber.from('2000'),
                    BigNumber.from('3000'),
                    BigNumber.from('4000'),
                    BigNumber.from('5000'),
                ],
            })
        ),

        getERC20Balance: jest.fn(() => Promise.resolve({ balance: 10000000 })),

        cancelLendingOrder: jest.fn(() =>
            Promise.resolve({
                wait: jest.fn(() => Promise.resolve({ blockNumber: 123 })),
            })
        ),

        getERC20TokenContractAddress: jest.fn(() =>
            Promise.resolve('0xEd4733fE7BAc4C2934F7e9CE4e0696b2169701D8')
        ),

        mintERC20Token: jest.fn(() =>
            Promise.resolve({
                wait: jest.fn(() => Promise.resolve({ blockNumber: 123 })),
            })
        ),

        getCurrencies: jest.fn(() =>
            Promise.resolve(
                getCurrencyMapAsList().map(currency => currency.symbol)
            )
        ),

        getProtocolDepositAmount: jest.fn(() =>
            Promise.resolve({
                ETH: BigNumber.from('100000000000000000000'), // 100 ETH
                EFIL: BigNumber.from('100000000000000000000000'), // 100 000 EFIL
                USDC: BigNumber.from('1000000000000'), // 1 000 000 USDC
                WBTC: BigNumber.from('1000000000000'), // 1000 BTC
            })
        ),

        unwindOrder: jest.fn(() =>
            Promise.resolve({
                hash: '0x123',
                wait: jest.fn(() =>
                    Promise.resolve({
                        blockNumber: 123,
                    })
                ),
            })
        ),

        getCollateralParameters: jest.fn(() =>
            Promise.resolve({
                liquidationThresholdRate: BigNumber.from('12500'),
            })
        ),

        getUsedCurrenciesForOrders: jest.fn(() =>
            Promise.resolve([ethBytes32, efilBytes32])
        ),

        getOrderList: jest.fn(() =>
            Promise.resolve({
                activeOrders: [
                    {
                        orderId: 1,
                        ccy: ethBytes32,
                        side: 0,
                        maturity: BigNumber.from(dec22Fixture.toString()),
                        unitPrice: BigNumber.from('9800'),
                        amount: BigNumber.from('1000000000'),
                        timestamp: BigNumber.from('1609295092'),
                    },
                    {
                        orderId: 2,
                        ccy: ethBytes32,
                        side: 0,
                        maturity: BigNumber.from(mar23Fixture.toString()),
                        unitPrice: BigNumber.from('9800'),
                        amount: BigNumber.from('1000000000'),
                        timestamp: BigNumber.from('1609295092'),
                    },
                    {
                        orderId: 1,
                        ccy: efilBytes32,
                        side: 1,
                        maturity: BigNumber.from(dec22Fixture.toString()),
                        unitPrice: BigNumber.from('9800'),
                        amount: BigNumber.from('1000000000'),
                        timestamp: BigNumber.from('1609295092'),
                    },
                ],
                inactiveOrders: [
                    {
                        orderId: 4,
                        ccy: ethBytes32,
                        side: 0,
                        maturity: BigNumber.from(dec22Fixture.toString()),
                        unitPrice: BigNumber.from('9800'),
                        amount: BigNumber.from('1000000000'),
                        timestamp: BigNumber.from('1609295092'),
                    },
                    {
                        orderId: 5,
                        ccy: efilBytes32,
                        side: 0,
                        maturity: BigNumber.from(dec22Fixture.toString()),
                        unitPrice: BigNumber.from('9800'),
                        amount: BigNumber.from('1000000000'),
                        timestamp: BigNumber.from('1609295092'),
                    },
                ],
            })
        ),
    };

    return mockSecuredFinance;
};
