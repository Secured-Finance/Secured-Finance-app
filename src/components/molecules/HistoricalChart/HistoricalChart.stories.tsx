import type { Meta, StoryFn } from '@storybook/react';
import { tradingData } from 'src/stories/mocks/historicalchart';
import { HistoricalChart } from '.';

const renderData = tradingData.map(item => {
    return {
        time: item[0],
        open: item[1],
        high: item[2],
        low: item[3],
        close: item[4],
        vol: item[5],
    };
});

export default {
    title: 'Molecules/HistoricalChart',
    component: HistoricalChart,
    args: {
        className: 'font-bold w-[800px]',
        data: renderData,
    },
} as Meta<typeof HistoricalChart>;

const Template: StoryFn<typeof HistoricalChart> = args => (
    <HistoricalChart {...args} />
);

export const Default = Template.bind({});
