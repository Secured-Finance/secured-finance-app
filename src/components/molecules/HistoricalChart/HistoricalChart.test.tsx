import { composeStories } from '@storybook/react';
import { mockUseSF } from 'src/stories/mocks/useSFMock';
import { render, screen, waitFor } from 'src/test-utils.js';
import * as stories from './HistoricalChart.stories';

const { Default } = composeStories(stories);

const mockSecuredFinance = mockUseSF();
jest.mock('src/hooks/useSecuredFinance', () => () => mockSecuredFinance);

describe('HistoricalChart component', () => {
    it('should render chart containers', () => {
        render(<Default />);

        expect(screen.getByTestId('candlestick-chart')).toBeInTheDocument();
        expect(screen.getByTestId('volume-chart')).toBeInTheDocument();
    });

    it('should render legend', () => {
        render(<Default />);

        expect(screen.getByText('Vol(USDT):')).toBeInTheDocument();
        expect(screen.getByText('O')).toBeInTheDocument();
        expect(screen.getByText('H')).toBeInTheDocument();
        expect(screen.getByText('L')).toBeInTheDocument();
        expect(screen.getByText('C')).toBeInTheDocument();
        expect(screen.getByText('Change')).toBeInTheDocument();
    });

    it('should call getHistoricalData with correct params when mounted', async () => {
        const { store } = await waitFor(() =>
            render(<Default />, {
                apolloMocks: Default.parameters?.apolloClient.mocks,
            })
        );
        expect(store.getState().landingOrderForm.maturity).toEqual(1669852800);
        expect(store.getState().landingOrderForm.currency).toEqual('WFIL');
    });

    it('should match snapshot', () => {
        const { container } = render(<Default />);
        expect(container).toMatchSnapshot();
    });
});
