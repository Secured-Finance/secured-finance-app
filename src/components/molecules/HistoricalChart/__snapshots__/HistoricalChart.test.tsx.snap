// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoricalChart component should match snapshot 1`] = `
<div>
  <div
    class="w-[600px]"
  >
    <div
      class="bg-neutral-900 pt-1 font-tertiary laptop:pt-2.5"
    >
      <div
        class="z-10 flex gap-4 text-2xs text-neutral-4 font-medium text-[11px] tablet:text-xs leading-4 pt-2 laptop:pt-1.5 px-4"
      >
        <div
          class="flex gap-1"
        >
          <span
            class="text-neutral-4"
          >
            O
          </span>
          <div
            class="font-normal text-palePurple"
          />
        </div>
        <div
          class="flex gap-1"
        >
          <span
            class="text-neutral-4"
          >
            H
          </span>
          <div
            class="font-normal text-palePurple"
          />
        </div>
        <div
          class="flex gap-1"
        >
          <span
            class="text-neutral-4"
          >
            L
          </span>
          <div
            class="font-normal text-palePurple"
          />
        </div>
        <div
          class="flex gap-1"
        >
          <span
            class="text-neutral-4"
          >
            C
          </span>
          <div
            class="font-normal text-palePurple"
          />
        </div>
        <div
          class="flex gap-1"
        >
          <span
            class="text-neutral-4 hidden laptop:block"
          >
            Change
          </span>
          <div
            class="font-normal text-nebulaTeal"
          />
        </div>
      </div>
      <div
        class="relative h-[156px] w-full laptop:h-[184px]"
        data-testid="candlestick-chart"
      >
        <div
          class="tv-lightweight-charts"
          style="overflow: hidden; direction: ltr; width: 0px; height: 260px; user-select: none; webkit-user-select: none; -webkit-tap-highlight-color: transparent;"
        >
          <table
            cellspacing="0"
            style="height: 260px; width: 0px;"
          >
            <tr>
              <td
                style="padding: 0px;"
              />
              <td
                style="padding: 0px; position: relative; width: 0px; height: 260px;"
              >
                <div
                  style="width: 100%; height: 100%; position: relative; overflow: hidden;"
                >
                  <canvas
                    style="width: 0px; height: 260px; position: absolute; z-index: 1; left: 0px; top: 0px; border-bottom: 2px solid #334155;"
                  />
                  <canvas
                    style="width: 0px; height: 260px; position: absolute; z-index: 2; left: 0px; top: 0px;"
                  />
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  style="height: 260px; overflow: hidden; width: 58px; left: 0px; position: relative;"
                >
                  <canvas
                    style="width: 58px; height: 260px; position: absolute; z-index: 1; left: 0px; top: 0px;"
                  />
                  <canvas
                    style="width: 58px; height: 260px; position: absolute; z-index: 2; left: 0px; top: 0px;"
                  />
                </div>
              </td>
            </tr>
            <tr
              style="display: none;"
            >
              <td
                style="padding: 0px;"
              />
              <td
                style="height: 25px; padding: 0px;"
              >
                <div
                  style="width: 100%; height: 100%; position: relative; overflow: hidden;"
                >
                  <canvas
                    style="width: 16px; height: 16px; position: absolute; z-index: 1; left: 0px; top: 0px;"
                  />
                  <canvas
                    style="width: 16px; height: 16px; position: absolute; z-index: 2; left: 0px; top: 0px;"
                  />
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  style="width: 25px; height: 100%; overflow: hidden;"
                >
                  <canvas
                    style="width: 16px; height: 16px;"
                  />
                </div>
              </td>
            </tr>
          </table>
        </div>
      </div>
      <div
        class="z-10 flex gap-4 text-2xs text-neutral-4 font-medium text-[11px] tablet:text-xs leading-4 pt-2 laptop:pt-1.5 px-4"
      >
        <div>
          <span>
            Vol(WFIL DEC22)
            :
          </span>
        </div>
        <div>
          <span>
            Vol(USDT):
          </span>
        </div>
      </div>
      <div
        class="relative h-[92px] w-full laptop:h-[115px]"
        data-testid="volume-chart"
      >
        <div
          class="tv-lightweight-charts"
          style="overflow: hidden; direction: ltr; width: 0px; height: 170px; user-select: none; webkit-user-select: none; -webkit-tap-highlight-color: transparent;"
        >
          <table
            cellspacing="0"
            style="height: 170px; width: 0px;"
          >
            <tr>
              <td
                style="padding: 0px;"
              />
              <td
                style="padding: 0px; position: relative; width: 0px; height: 142px;"
              >
                <div
                  style="width: 100%; height: 100%; position: relative; overflow: hidden;"
                >
                  <canvas
                    style="width: 0px; height: 142px; position: absolute; z-index: 1; left: 0px; top: 0px;"
                  />
                  <canvas
                    style="width: 0px; height: 142px; position: absolute; z-index: 2; left: 0px; top: 0px;"
                  />
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  style="height: 142px; overflow: hidden; width: 58px; left: 0px; position: relative;"
                >
                  <canvas
                    style="width: 58px; height: 142px; position: absolute; z-index: 1; left: 0px; top: 0px;"
                  />
                  <canvas
                    style="width: 58px; height: 142px; position: absolute; z-index: 2; left: 0px; top: 0px;"
                  />
                </div>
              </td>
            </tr>
            <tr>
              <td
                style="padding: 0px;"
              />
              <td
                style="height: 28px; padding: 0px; width: 0px;"
              >
                <div
                  style="width: 100%; height: 100%; position: relative; overflow: hidden;"
                >
                  <canvas
                    style="width: 0px; height: 28px; position: absolute; z-index: 1; left: 0px; top: 0px;"
                  />
                  <canvas
                    style="width: 0px; height: 28px; position: absolute; z-index: 2; left: 0px; top: 0px;"
                  />
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  style="width: 58px; height: 28px; overflow: hidden;"
                >
                  <canvas
                    height="28"
                    style="width: 58px; height: 28px;"
                    width="58"
                  />
                </div>
              </td>
            </tr>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
